server/src/main/java/org/elasticsearch/action/admin/indices/stats/TransportDataStreamsStatsAction.java

```diff
@@ -0,0 +1,304 @@
+        if (dateMathExpression) {
+            assert IndexMetadata.State.values().length == 2 : "a new IndexMetadata.State value may need to be handled!";
+            return true;
+        }
+        if (indexMetadata.getState() == IndexMetadata.State.CLOSE && indicesOptions.expandWildcardsClosed()) {
+            return true;
+        }
+        if (indexMetadata.getState() == IndexMetadata.State.OPEN && indicesOptions.expandWildcardsOpen()) {
+            return true;
+        }
+        return false;
+    }
+
+    private static boolean isVisibleDueToImplicitHidden(String expression, String index) {
+        return index.startsWith(".") && expression.startsWith(".") && Regex.isSimpleMatchPattern(expression);
+    }
+
+    @Override
+    protected ShardsIterator shards(ClusterState clusterState, DataStreamsStatsRequest request, String[] concreteIndices) {
+        String[] requestIndices = request.indices();
+        if (requestIndices == null || requestIndices.length == 0) {
+            requestIndices = new String[]{"*"};
+        }
+        List<String> abstractionNames = resolveIndexAbstractions(requestIndices, request.indicesOptions(), clusterState.getMetadata(), indexNameExpressionResolver);
+        SortedMap<String, IndexAbstraction> indicesLookup = clusterState.getMetadata().getIndicesLookup();
+
+        String[] concreteDatastreamIndices = abstractionNames.stream().flatMap(abstractionName -> {
+            IndexAbstraction indexAbstraction = indicesLookup.get(abstractionName);
+            assert indexAbstraction != null;
+            if (indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM) {
+                IndexAbstraction.DataStream dataStream = (IndexAbstraction.DataStream) indexAbstraction;
+                List<IndexMetadata> indices = dataStream.getIndices();
+                return indices.stream().map(idx -> idx.getIndex().getName());
+            } else {
+                return Stream.empty();
+            }
+        }).toArray(String[]::new);
+        return clusterState.getRoutingTable().allShards(concreteDatastreamIndices);
+    }
+
+    @Override
+    protected DataStreamShardStats shardOperation(DataStreamsStatsRequest request, ShardRouting shardRouting) throws IOException {
+        IndexService indexService = indicesService.indexServiceSafe(shardRouting.shardId().getIndex());
+        IndexShard indexShard = indexService.getShard(shardRouting.shardId().id());
+        // if we don't have the routing entry yet, we need it stats wise, we treat it as if the shard is not ready yet
+        if (indexShard.routingEntry() == null) {
+            throw new ShardNotFoundException(indexShard.shardId());
+        }
+        StoreStats storeStats = indexShard.storeStats();
+        return new DataStreamShardStats(
+            indexShard.routingEntry(),
+            storeStats,
+            0L //TODO
```

Comment:
Working on doing the lucene searching at the shard level to obtain this. Could use some guidance here.

Reply:
I think something like this would work here:

```java
IndexAbstraction indexAbstraction = 
   clusterService.state().metadata().getIndicesLookup().get(shardRouting.getIndexName());
DataStream dataStream = indexAbstraction.getParentDataStream().getDataStream();
try (Engine.Searcher engineSearcher = indexShard.acquireSearcher("data_stream_stats")) {
  IndexReader indexReader = engineSearcher.getIndexReader();
  String fieldName = dataStream.getTimeStampField().getName(); // Should be the fieldname in the Lucene index
  long maxValue = LongPoint.decodeDimension(PointValues.getMaxPackedValue(indexReader, fieldName), 0); // dates (and numbers) are stored as point fields (backed by a bkd tree) and we can efficiently get the max value per segment and this method gets the max for all segments per Lucene index 
  // indexShard.mapperService().fieldType(fieldName).valueForDisplay(); perhaps format to a printable name instead of seconds since epoch? 
 return new DataStreamShardStats(
    indexShard.routingEntry(),
    storeStats,
    maxValue
  );
}
```